stats/stats.go

```diff
@@ -198,24 +208,53 @@ func NewHistogram(Offset float64, Divider float64) *Histogram {
 	return h
 }
 
-// Tradeoff memory for speed (though that also kills the cache so...)
-// this creates an array of 100k (max value) entries
-// TODO: consider using an interval search for the last N big buckets
+// Val2Bucket values are kept in two different structure
+// constantTimeVal2Bucket allows you reach between 0 and 1000 in constant time
+// linearTimeVal2Bucket allows you to reach between 1000 and 100000 in linear time
+// linearTimeVal2Bucket approach saves memory usage but loses speed
 func init() {
 	lastV := int32(lastValue)
-	val2Bucket = make([]int, lastV+1)
+	line5EndValue := histogramBucketValues[line5EndIndex]
+	constantTimeVal2Bucket = make([]int, line5EndValue)
+	linearTimeVal2Bucket = make([]UpperValueIndexer, line5EndToEndLength)
+	val2BucketLen := int32(len(constantTimeVal2Bucket))
 	idx := 0
-	for i := int32(0); i <= lastV; i++ {
+	for i := int32(0); i < val2BucketLen; i++ {
 		if i >= histogramBucketValues[idx] {
 			idx++
 		}
-		val2Bucket[i] = idx
+		constantTimeVal2Bucket[i] = idx
 	}
+	idx++ // Not include 1000
+	j := 0
+	for ; j < line5EndToEndLength-1; j++ {
+		currentValue := histogramBucketValues[idx]
+		previosValue := histogramBucketValues[idx-1]
+		linearTimeVal2Bucket[j] = UpperValueIndexer{Index: idx, Range: int(currentValue - previosValue)}
+		idx++
+	}
+	linearTimeVal2Bucket[j] = UpperValueIndexer{Index: idx, Range: 1} // for 100000
 	// coding bug detection (aka impossible if it works once)
 	if idx != numValues {
 		log.Fatalf("Bug in creating histogram buckets idx %d vs numbuckets %d (last val %d)", idx, numValues, lastV)
 	}
+}
 
+// lookUpIdx looks for scaledValue's index in histogramBucketValues
+func lookUpIdx(scaledValue int) int {
+	constantOrLinearDeciderValue := int(histogramBucketValues[line5EndIndex])
+	if constantOrLinearDeciderValue > scaledValue { //constant
```

Comment:
super minor nit but it would read (to me) easier if it was
if scaledValue <= constantOrLinearDeciderValue

also maybe constantOrLinearDeciderValue should just be maxLookupValue or some such